annnufan/assignment4/islands_counter.cpp

```diff
@@ -0,0 +1,59 @@
+#include <vector>
+#include <utility>
+
+class islands_counter {
+	std::vector<std::vector<bool>> island_map;
+	std::vector<std::pair<int, int>> moves;
+	int n, m;
+	int islands;
+	bool is_count;
+
+	void dfs(int i, int j) {
+		island_map[i][j] = false;
+
+		for (const auto& move : moves) {
+			int i1 = move.first + i;
+			int j1 = move.second + j;
+			if (0 <= i1 && i1 < n && 0 <= j1 && j1 < m && island_map[i1][j1]) {
+				dfs(i1, j1);

```

Comment:
My personal opinion:
I prefer names that are less "mathematical", and more related to the context.
I think `DFS` is a good name when accepting a general graph, and performing a general, mathematical DFS algorithm.
Here this DFS actually have a purpose: it markes islands as visited (more like - removes them from the map).
So I think `mark_island` or `remove_island` or `clear_island` are better names.
Then with a comment, you can specify that you implemented it with DFS (as it is an implementation detail).

This matches the pattern in which the function/class name represents its interface and not its implemntation, represents the WHAT and not the HOW.
For example, this is why in C++, a linked list is called "list", a tree is called "set" and a hash table is called "unordered_set". The 'hash table' is just an implementation detail; the actual interface is a set that has not order on the elements. 
This is from a pure "object oriented" perspective - encapsulate!

Arguments against what I just said:
1. This is a private function so you can have freedom in how you call it. This is not as important as the outer interface of the public functions
2. implementation is still important sometimes and you do not always want to completely encapsulate it. For example, everyone should know that the unordered_set is a hash table and understand the implications of it. So keep that in mind as well. The "pure OOP" is not always the right way to go, though it it probably best most of the time.